core/unit_test/TestNumericTraits.hpp

```diff
@@ -0,0 +1,327 @@
+  DEFINE_EXTREMA(unsigned short, 0, USHRT_MAX);
+  DEFINE_EXTREMA(int, INT_MIN, INT_MAX);
+  DEFINE_EXTREMA(unsigned, 0U, UINT_MAX);
+  DEFINE_EXTREMA(long, LONG_MIN, LONG_MAX);
+  DEFINE_EXTREMA(unsigned long, 0UL, ULONG_MAX);
+  DEFINE_EXTREMA(long long, LLONG_MIN, LLONG_MAX);
+  DEFINE_EXTREMA(unsigned long long, 0ULL, ULLONG_MAX);
+
+  DEFINE_EXTREMA(float, -FLT_MAX, FLT_MAX);
+  DEFINE_EXTREMA(double, -DBL_MAX, DBL_MAX);
+  DEFINE_EXTREMA(long double, -LDBL_MAX, LDBL_MAX);
+
+#undef DEFINE_EXTREMA
+};
+
+// clang-format off
+struct Infinity { template <class T> using trait = Kokkos::Experimental::infinity<T>; };
+struct Epsilon { template <class T> using trait = Kokkos::Experimental::epsilon<T>; };
+struct FiniteMin { template <class T> using trait = Kokkos::Experimental::finite_min<T>; };
+struct FiniteMax { template <class T> using trait = Kokkos::Experimental::finite_max<T>; };
+struct RoundError { template <class T> using trait = Kokkos::Experimental::round_error<T>; };
+struct NormMin { template <class T> using trait = Kokkos::Experimental::norm_min<T>; };
+struct Digits { template <class T> using trait = Kokkos::Experimental::digits<T>; };
+struct Digits10 { template <class T> using trait = Kokkos::Experimental::digits10<T>; };
+struct MaxDigits10 { template <class T> using trait = Kokkos::Experimental::max_digits10<T>; };
+struct Radix { template <class T> using trait = Kokkos::Experimental::radix<T>; };
+struct MinExponent { template <class T> using trait = Kokkos::Experimental::min_exponent<T>; };
+struct MaxExponent { template <class T> using trait = Kokkos::Experimental::max_exponent<T>; };
+struct MinExponent10 { template <class T> using trait = Kokkos::Experimental::min_exponent10<T>; };
+struct MaxExponent10 { template <class T> using trait = Kokkos::Experimental::max_exponent10<T>; };
+// clang-format on
+
+template <class T>
+KOKKOS_FUNCTION T* take_address_of(T& arg) {
+  return &arg;
+}
+
+template <class T>
+KOKKOS_FUNCTION void take_by_value(T) {}
+
+template <class Space, class T, class Tag>
+struct TestNumericTraits {
+  template <class U>
+  using trait = typename Tag::template trait<U>;
+
+  TestNumericTraits() { run(); }
+
+  void run() const {
+    int errors = 0;
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<Space, Tag>(0, 1), *this,
+                            errors);
+    ASSERT_EQ(errors, 0);
+    (void)take_address_of(trait<T>::value);  // use on host
+  }
+
+  KOKKOS_FUNCTION void operator()(Infinity, int, int& e) const {
+    using Kokkos::Experimental::infinity;
+    auto const inf  = infinity<T>::value;
+    auto const zero = T(0);
+    e += (int)!(inf + inf == inf);
+    e += (int)!(inf != zero);
+    use_on_device();
+  }
+
+  KOKKOS_FUNCTION void operator()(Epsilon, int, int& e) const {
+    using Kokkos::Experimental::epsilon;
+    auto const eps = epsilon<T>::value;
+    auto const one = T(1);
+    e += (int)!(one + eps != one);
+    e += (int)!(one + eps / 2 == one);
```

Comment:
The Fedora Intel builds seem to not like this line

Reply:
Fixed in 5913925.
For reference https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/floating-point-options/fp-model-fp.html
`-fp-model source` would also done the trick.